### PR TITLE
Cleaner iterative N0 N1 library

### PR DESCRIPTION
**Pull request by louisl3grand**
I have modified the iterative N0 N1 library, so that it is a bit cleaner. It also works with the TT and MV estimators now. 
You can specify a lib_dir to cache the stuff, and there is a hashcheck to verify that the parameters in cache are the same. 

In the function `get_n0n1`, when the N0 and N1 bias are not estimated with the fiducial Cls, the cacher will give it a name that contains the hash key of the true Cls unl. 

I put the class in the file `iterbiasesN0N1.py` as it breaks the functions in the previous `n0n1_iterative.py`, but the idea is to replace the later.

### CODE REVIEW

File: delensalot/biases/iterbiasesN0N1.py [hunk 1]
```diff
@@ -0,0 +1,424 @@
+# Adapted from Julien Carron lenspec script
+
+import os
+import numpy as np
+import hashlib
+import pickle as pk
+from healpy import gauss_beam
+from scipy.interpolate import UnivariateSpline as spl
+from plancklens.helpers import cachers, mpi
```
Comment (Sebastian-Belkner) [inline delensalot/biases/iterbiasesN0N1.py:9]: delensalot has its own mpi and cachers, please use them instead. Also plancklens.mpi is turned off by default when run via config.

(In the near future we want to remove plancklens dependencies)

File: delensalot/biases/iterbiasesN0N1.py [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+
+        rho.append(rho_sqd_phi)
+        delcls_fid.append(cls_plen_fid)
+        delcls_true.append(cls_plen_true)
+
+    return delcls_fid, delcls_true
+
+
+
+def cls2N0N1(k, cls_cmb_filt, cls_cmb_dat, cls_noise_filt, cls_noise_dat, lmin_ivf, lmax_ivf, lmax_qlm, doN1mat=False):
```
Comment (Sebastian-Belkner) [inline delensalot/biases/iterbiasesN0N1.py:376]: Some type annotation for this and all other functions would be great. Also docstring for some parameter missing, e.g. what is k? i think it is the same as qe_key in function above, can we uniformize the naming please?

File: delensalot/biases/iterbiasesN0N1.py [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+        return fid_delcls, true_delcls
+
+
+def get_fals(qe_key, cls_cmb_filt, cls_cmb_dat, cls_noise_filt, cls_noise_dat, lmin_ivf, lmax_ivf):
+    """
+    Get the filtering Cls and from the fiducial CMB Cls and noise, as well as the IVF data Cls
+    Returns as well the QE weights and CMB response functions 
+    
+    Args:
+        qe_key: 'ptt', 'p_p', or 'p'
```
Comment (Sebastian-Belkner) [inline delensalot/biases/iterbiasesN0N1.py:202]: We probably need an assert here to catch other keys this function does not support?

**Review by Sebastian-Belkner — CHANGES_REQUESTED** [review]
Hi,
I think this piece of code is very important to delensalot and will be used heavily in the future. It replaces the old implementation and seems easier to read and use.

In addition to the other comments:
Please provide an example scripts to see how the functions are used and what to expect from them. Once provided, I can write a test so we can be always sure that the N0 and N1 predictions work as intended.

File: delensalot/biases/iterbiasesN0N1.py [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+from plancklens.helpers import cachers, mpi
+from plancklens import qresp, nhl, utils
+
+#Uses lensitbiases to compute fast N1 
+from lensitbiases import n1_fft
+
+# Requires the full camb python package for the lensed spectra calc.
+from camb.correlations import lensed_cls
+
+def cls2dls(cls):
```
Comment (Sebastian-Belkner) [inline delensalot/biases/iterbiasesN0N1.py:18]: This and the inverse is already part of [utils](https://github.com/NextGenCMB/delensalot/blob/main/delensalot/utils.py)

**Comment by louisl3grand** [comment]
I have taken into account your comments and I put a demo notebooks in first_steps/notebooks/non1_iter.ipynb

File: delensalot/biases/iterbiasesN0N1.py (hunk 1 above)
Comment (louisl3grand) [reply delensalot/biases/iterbiasesN0N1.py]: There would be some functions that come from plancklens that would be redundant if we copy paste them in delensalot, such as qresp.py and nhl.py

**Review by Sebastian-Belkner — APPROVED** [review]
Looks good!

**Review by Sebastian-Belkner — APPROVED** [review]
Looks good!

**Pull request merged**